{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cd473a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readGenome(filename):\n",
    "    genome = \"\"\n",
    "    with open(filename,'r') as f:\n",
    "        for line in f:\n",
    "            if not line[0] == \">\":\n",
    "                genome += line.rstrip()\n",
    "    return genome\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00984d1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:1: SyntaxWarning: invalid escape sequence '\\C'\n",
      "<>:1: SyntaxWarning: invalid escape sequence '\\C'\n",
      "C:\\Users\\hp\\AppData\\Local\\Temp\\ipykernel_2924\\873311507.py:1: SyntaxWarning: invalid escape sequence '\\C'\n",
      "  genome = readGenome(\"F:\\Coursera\\Algorithms for DNA sequences\\lambda_virus.fa\")\n"
     ]
    }
   ],
   "source": [
    "genome = readGenome(\"F:\\Coursera\\Algorithms for DNA sequences\\lambda_virus.fa\") #Add your file path as an argument\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c8595757",
   "metadata": {},
   "outputs": [],
   "source": [
    "def naive(p,t):\n",
    "    \"\"\"Matches the occurences of pattern 'p' in text 't'.\"\"\"\n",
    "    occurences = []\n",
    "    for i in range(len(t)-len(p)+1):\n",
    "        match = True\n",
    "        for j in range(len(p)):\n",
    "            if not t[i+j] == p[j]:\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurences.append(i)\n",
    "    return occurences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf31ace1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "215\n"
     ]
    }
   ],
   "source": [
    "def naive_2mm(p,t, distance):\n",
    "    \"\"\"Matches the occurences of pattern 'p' in text 't' with defined mismatches.\"\"\"\n",
    "    occurences = []\n",
    "    for i in range(len(t)-len(p)+1):\n",
    "        match = True\n",
    "        mismatch = 0\n",
    "        for j in range(len(p)):\n",
    "            if t[i+j] == p[j]:\n",
    "                match\n",
    "            if not t[i+j] == p[j]:\n",   
    "                mismatch += 1\n",
    "            if mismatch > distance: #Compare number of mismatch with distance.\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurences.append(i)\n",
    "    return occurences\n",
    "total = naive_2mm(\"AGGAGGTT\", genome, distance) #Example of counting the number of occurences of a pattern in a genome\n",
    "print(len(total))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8d969fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "def generateReads(genome, numreads, readlen):\n",
    "    \"\"\"Generate reads from a random positions within genome\"\"\"\n",
    "    reads = []\n",
    "    for _ in range(numreads):\n",
    "        start = random.randint(0, len(genome)-readlen) - 1\n",
    "        reads.append(genome[start:start+ readlen])\n",
    "    return reads\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d481791",
   "metadata": {},
   "outputs": [],
   "source": [
    "reads = generateReads(genome, 100, 100)\n",
    "\"\"\"Checking the naive function with randomly created reads\"\"\"\n",
    "\n",
    "numMatched = 0\n",
    "for r in reads:\n",
    "    matches = naive(r, genome)      #returns list('matches') that containes the occureneces of reads in genome\n",
    "    if len(matches)> 0:\n",
    "        numMatched += 1             #counts the number of matches of a particular read at place against the genome.\n",
    "print(\"%d / %d reads matched!\" % (numMatched, len(reads)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "911ce434",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Readfastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as f:\n",
    "        while True:\n",
    "            f.readline()                #ignore first line of a read i.e., Name\n",
    "            seq = f.readline().rstrip() # removes white space while reading only add sequence\n",
    "            f.readline()                #ignore third line of a read i.e., placeholder strand (+ or -)\n",
    "            qual = f.readline().rstrip() # removes white space while reading only add ASCII encoded Quality Value\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f238f079",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_reads, _ = Readfastq(\"F:\\Coursera\\Algorithms for DNA sequences\\ERR266411_1.first1000.fastq\")  #Add your file path as an argument\n",
    "\"\"\"Performing the naivev function with actual seq_reads\"\"\"\n",
    "\n",
    "numMatched = 0\n",
    "n = 0       # to keep track of how many reads were processed\n",
    "for r in seq_reads:\n",
    "    \n",
    "    matches = naive(r, genome)\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "\n",
    "print(\"%d / %d reads matched against the genome!\" % (numMatched, n))\n",
    "# The code above results in a very few matches of reads against the genome even though the Fastq reads and genome are of the same species.\n",
    "# This might be due to error of sequencing machine.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c248faf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The code below tries to match only the portion of read against the genome to obtain the expected higher.\n",
    "\"\"\"Performing the naivev function with actual seq_reads\"\"\"\n",
    "\n",
    "numMatched = 0\n",
    "n = 0       # to keep track of how many reads were processed\n",
    "for r in seq_reads:\n",
    "    r = r[:30]\n",
    "    matches = naive(r, genome)\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "\n",
    "print(\"%d / %d reads matched against the genome!\" % (numMatched, n))\n",
    "\n",
    "# The output matches are significantly increased though it is only less than half of the genome.\n",
    "# This is because the genome is double stranded while particular reads come from either strand of the genome.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6a9b3ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverseComplement(S):\n",
    "    complement = {\"A\":\"T\",\"T\":\"A\",\"C\":\"G\",\"G\":\"C\", \"N\":\"N\"} # Here, 'N' is the nucleotide in a read for which sequencer is not confident enough to make a call between any four nucleotide\n",
    "    t = \"\"\n",
    "    for base in S:\n",
    "        t = complement[base] + t\n",
    "    return t\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66886f08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code below compares the reads as well as reversecomplement of reads against the genome.\n",
    "\"\"\"Performing the naivev function with actual seq_reads\"\"\"\n",
    "\n",
    "numMatched = 0\n",
    "n = 0       # to keep track of how many reads were processed\n",
    "for r in seq_reads:\n",
    "    r = r[:30]\n",
    "    reverse_r = reverseComplement(r)    \n",
    "    matches = naive(r, genome) # Returns the list of matches of forward read against the genome\n",
    "    if not reverse_r == r:\n",
    "        matches_r = naive(reverse_r, genome)\n",
    "        matches.extend(matches_r) # extend the above list by adding the lists containing the matches of reverse complement of that read against the genome\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "        \n",
    "\n",
    "print(\"%d / %d reads matched against the genome!\" % (numMatched, n))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
